helpers/traits/AltapayMaster.php

```diff
@@ -243,4 +245,60 @@ public function getLatestTransaction( $transactions, $authType ) {
 
 		return $latest_transaction;
 	}
+
+	/**
+	 * @param $order_id
+	 * @param $txn_id
+	 * @param $transaction
+	 * @param $fraud_recommendation
+	 *
+	 * @return bool
+	 */
+	public function detectFraud( $order_id, $txn_id, $transaction, $fraud_recommendation ) {
+		$return             = false;
+		$detect_fraud       = get_option( 'altapay_fraud_detection' );
+		$do_action_on_fraud = get_option( 'altapay_fraud_detection_action' );
+		if ( $detect_fraud and $do_action_on_fraud and $fraud_recommendation === 'Deny' ) {
+			$return = true;
+			try {
+				$auth = $this->getAuth();
+				if ( $transaction['TransactionStatus'] === 'captured' and ! get_post_meta( $order_id, '_refunded', true ) ) {
+					$reconciliation_id = wp_generate_uuid4();
+					$api               = new RefundCapturedReservation( $auth );
+					$api->setReconciliationIdentifier( $reconciliation_id );
+				} elseif ( get_post_meta( $order_id, '_released', true ) ) {
+					$api = new ReleaseReservation( $auth );
+				}
+				$api->setTransaction( $transaction['TransactionId'] );
+				$response = $api->call();
+				if ( $response->Result === 'Success' ) {
+					if ( ! empty( $reconciliation_id ) ) {
+						$transaction = json_decode( wp_json_encode( $response->Transactions ), true );
+						$transaction = reset( $transaction );
+
+						$reconciliation = new Core\AltapayReconciliation();
+						$reconciliation->saveReconciliationIdentifier( (int) $order_id, $transaction['TransactionId'], $reconciliation_id, 'refunded' );
+						update_post_meta( $order_id, '_refunded', true );
+
+						// Release agreement
+						if ( $txn_id != $transaction['TransactionId'] ) {
+							$api = new ReleaseReservation( $auth );
+							$api->setTransaction( $txn_id );
+							$response = $api->call();
+							if ( $response->Result !== 'Success' ) {
+								error_log( "altapay_fraud_detection_action error releasing agreement: $response->MerchantErrorMessage" );
+							}
+						}
+					} else {
+						update_post_meta( $order_id, '_released', true );
+					}
+				} else {
+					error_log( "altapay_fraud_detection_action error: $response->MerchantErrorMessage" );
+				}
+			} catch ( Exception $e ) {
+				error_log( "altapay_fraud_detection_action exception: {$e->getMessage()}" );
+			}
+		}
+		return $return;
```

Comment:
this piece of logic looks pretty complex. do we have any tests for it?

Reply:
This is a newly functionality, so, we do not have an automated test for it yet. But @taimurmansur has manually tested it.